brightnessctl.1
.TH "BRIGHTNESSCTL" "1" "2025-09-20" "brightnessctl" "brightnessctl"
. 
.SH "NAME"
brightnessctl \- read and control device brightness
.
.
.SH "SYNOPSIS"
\fBbrightnessctl\fR [\fBget\fR|\fBset\fR] [\fIOPTION\fR]... [\fIVALUE\fR]
.
.
.SH "DESCRIPTION"
.P
Controls and retrieves the brightness of various devices like backlights
or leds.
Depending on the compilation, it uses the dbus or writes to the
files in /sys/class directly, whereas the latter requires root to execute.
.
.
.SH "OPTIONS"
.SS "General"
.TP
.B \-m\fR, \fB\-\-machine-readable
produce machine readable output
.
.TP
.B \-d\fR, \fB\-\-device=\fIDEVICE
specify device name
.
.TP
.B \-c\fR, \fB\-\-class=\fICLASS
specify device class
.
.TP
.B \-h\fR, \fB\-\-help
print a help message
.
.TP
.B \-v\fR, \fB\-\-version
print version and exit
.
.TP
.B \-e\fR, \fB\-\-exponent\fR[\fB=\fIK\fR]
set exponent of brightnesscurve
.br
If the flag is not provided a linear curve is used (\fIK\fR=1).
If the flag is provided without any argument, \fIK\fR=3 is used.
.
.TP
.B \-q\fR, \fB\-\-quiet
suppress output
.br
This silences some error messages (but not error return codes) and
superflous messages for `get` and `set`.
.
.TP
.B \-s\fR, \fB\-\-save
save previous state in a temporary file
.
.
.SS "Set"
.P
The set operation sets the brightness of the device specified through
\fB\-\-device=\fIDEVICE\fR and \fB\-\-class=\fICLASS\fR.
In general, it expects a \fIVALUE\fR to be provided in one of the
following formats:
.
.TP
.I NN
where \fINN\fR is a non-negative integer
.br
This sets the brightness to the absolute value \fINN\fR.
.
.TP
.I NN\fB%
This sets the brightness to \fINN\fR% of the maximum brightness.
It will respect the \fB\-\-exponent\fR[\fB=\fIK\fR] setting.
The curve is:
.br
.I NN\fR%^\fIK\fR * MAXIMUM_BRIGHTNESS.
.
.TP
.B +\fINN\fR[\fB%\fR]
This increases the brightness by \fINN\fR[\fB%\fR].
.
.TP
.I NN\fR[\fB%\fR]\fB-
This decreases the brightness by \fINN\fR[\fB%\fR].
.
.P
The following are a list of options specific to the use with set.
.
.TP
.B \-n\fR, \fB\-\-min\-value=\fIn
set minimum brightness
.br
This overrides any value computation.
.
.TP
.B \-r\fR, \fB\-\-restore
restore previously saved state
.br
If this flag is set, providing a \fIVALUE\fR is not required as the
brightness will be restored from a previous call with \fB\-\-save\fR.
The two can be used together (after an initial save) to toggle between the
current and the previous state.
.
.SS "Get"
.P
The get operation retrieves the brightness of the specified device.
.
.TP
.B \-l\fR, \fB\-\-list
List info for all devices
.br
This is useful to get the class and name of a device if you don't know it.
.
.
.SH "EXIT STATUS"
.TP
0
everything is fine
.TP
1
an error occurred
.br
In some cases you will still get the output you wanted from this.
For example, if you do not (fully) specify the device you want to use,
the program will default to what is probably the backlight of your main
monitor.
When doing so it will also print a huge complaint about that if it is
not silenced with \fB\-\-quiet\fR.


src/brightnessctl.c
#include "brightnessctl.h"

#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

static struct options options = {
  .quiet = false,
  .machine_readable = false,
  .help = false,
  .version = false,
  .save = false,
  .restore = false,
  .list = false,
  .min_value = 0,
  .exponent = 1,
  .set_to = NULL,
  .operation = UNSET,
  .device = NULL,
  .class = NULL
};

int main(int argc, char *const *argv) {
  int opt;

  const char *optstring = "+qmd:c:hvsn:e::rl";

  const struct option long_options[] = {
    {"quiet", no_argument, NULL, 'q'},
    {"machine-readable", no_argument, NULL, 'm'},
    {"device", required_argument, NULL, 'd'},
    {"class", required_argument, NULL, 'c'},
    {"help", no_argument, NULL, 'h'},
    {"version", no_argument, NULL, 'v'},
    {"save", no_argument, NULL, 's'},
    // `set` specific arguments
    {"min-value", required_argument, NULL, 'n'},
    {"exponent", optional_argument, NULL, 'e'},
    {"restore", no_argument, NULL, 'r'},
    {"list", no_argument, NULL, 'l'},
    {NULL, 0, NULL, 0}
  };

  // Options start after the verb/operation
  char *const *opts = argv+1;
  int optc = argc-1;

  // Before argument processing, we check, whether a verb/operation was provided
  // otherwise, we send them to help
  if (argc >= 2) {
    if (!strcmp(argv[1], "get")) { // strcmp returns *0* if strings are equal
      options.operation = GET;
    } else if (!strcmp(argv[1], "set")) {
      options.operation = SET;
    } else {
      // The verb/operation was omitted
      // This can either be a user error, or --help or --version are invoked
      opts = argv;
      optc = argc;
    }
  }

  int error = 0;

  while ((opt = getopt_long(optc, opts, optstring, long_options, NULL)) != -1) {
    switch (opt) {
      case 'q':
        options.quiet = true;
        break;
      case 'm':
        options.machine_readable = true;
        break;
      case 'd':
        options.device = optarg;
        break;
      case 'c':
        options.class = optarg;
        break;
      case 'h':
        options.help = true;
        break;
      case 'v':
        options.version = true;
        break;
      case 's':
        options.save = true;
        break;
      case 'n':
        // Unchecked conversion from long to int
        // (User may not provide a value that is too high;
        // if they do, their minimum is not respected)
        options.min_value = strtol(optarg, NULL, 10);
        break;
      case 'e':
        // Unchecked conversion from long to int
        // (Very high exponents lead to unexpected behavior)
        if (optarg)
          options.exponent = strtol(optarg, NULL, 10);
        else
          options.exponent = 3;
        break;
      case 'r':
        options.restore = true;
        break;
      case 'l':
        options.list = true;
        break;
      default:
        // Something is really off
        error = 1;
        // Fallthrough
      case '?':
        // Problem with parsing `optopt` could provide insights
        // `getopt` prints error message on its own (probably?)
        options.help = true;
    }
  }

  // If applicable (and present) we get the value for the set command
  if (options.operation == SET) {
    if (optind < optc)
      options.set_to = opts[optind];
    else
      options.help |= !options.restore;
  } 

  // Now we determine what to do
  if (error) {
    // The parser behaved in an unexpected manner
    // Did we have a typo in an option or
    // did we forget to handle one?
    fprintf(
      stderr,
      "An unhandled error occured during argument parsing.\n"
      "This should never happen."
    );
  }
  if (options.help || error || (options.operation == UNSET && !options.version)) {
    print_help(&options);
    return error;
  }
  if (options.version) {
    print_version();
    return 0;
  }
  if (options.operation == GET) {
    return get(&options);
  }
  if (options.operation == SET) {
    return set(&options);
  }
  // If we reach this point something went wrong
  printf("Something unexpected happened.");
  return 1;
}
